ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -10,6 +10,43 @@ use crate::{
     },
 };
 
+/// This function computes the shares that sum to zero from the zero-knowledge proofs.
+pub fn compute_g_differences<F, const P: usize, const λ: usize>(
+    zkp: &Vec<[F; P]>,
+    challenges: &[F],
+) -> Vec<F>
+where
+    F: PrimeField,
+{
+    // compute denominator
+    let lagrange_denominator: CanonicalLagrangeDenominator<F, P> =
+        CanonicalLagrangeDenominator::<F, P>::new();
+
+    // compute g_r with "empty" first spot
+    let g_r = iter::once(F::ZERO)
+        .chain(
+            challenges
+                .iter()
+                .zip(zkp)
+                .map(|(challenge, zkp)| interpolate_at_r(zkp, *challenge, &lagrange_denominator)),
+        )
+        .collect::<Vec<_>>();
+
+    // compute g_sum)
+    let g_sum = zkp
+        .iter()
+        .map(compute_sum_share::<F, λ, P>)
+        // append spot for final sum
+        .chain(iter::once(F::ZERO))
+        .collect::<Vec<_>>();
+
+    g_sum
```

Comment:
Instead of doing this in 3 stages, i suggest we just combine them into a single map, that just computes the differences directly.

Reply:
I could use a single collect, i.e. not collecting g_r and g_sum in a vector (which would make the code more complex). I am not sure how to use a single map, the challenge is that for the first and last element I need to do something different which is hard to do with a single map (unless we enumerate the iterator and put an if statement in the map).